File: benchmarks/comparable/python/map_build.py
# map_build.py - CPython/protoPython twin of benchmarks/comparable/map_build.scala:
# build a 50000-entry dict keyed by str, read every key back, and fold the values.
# Result: 50000 1249975000 50000 (computed with tools/scala3-3.9.0).
#
# Python's dict is mutable and Scala's Map is not, so the twin measures the same
# algorithm rather than the same allocation behaviour; benchmarks/README.md says
# so for every workload whose data structure differs.


def main():
    m = {}
    i = 0
    while i < 50000:
        m["k" + str(i)] = i
        i += 1
    total = 0
    found = 0
    j = 0
    while j < 50000:
        v = m.get("k" + str(j), -1)
        if v >= 0:
            total += v
            found += 1
        j += 1
    print(str(len(m)) + " " + str(total) + " " + str(found))


main()

File: benchmarks/comparable/map_build.scala
// EXPECT: 50000 1249975000 50000
// map_build.scala - build a 50000-entry Map keyed by String, read every key
// back, and fold the values. The ROADMAP's benchmark suite v1 "map-build"
// workload (Phase 3). It prints the entry count, the value fold and the number
// of keys read back successfully, so a Map that loses entries fails the run.
// The three numbers were computed with tools/scala3-3.9.0, never by hand.

@main def benchMapBuild(): Unit =
  var m = Map[String, Int]()
  var i = 0
  while i < 50000 do
    m = m + (("k" + i) -> i)
    i += 1
  var total = 0
  var found = 0
  var j = 0
  while j < 50000 do
    val v = m.getOrElse("k" + j, -1)
    if v >= 0 then
      total += v
      found += 1
    j += 1
  println(m.size.toString + " " + total + " " + found)
